File: backend/app/api/v1/endpoints/scenarios.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from app.db.session import get_db
from app.schemas.scenario import ScenarioCreate, ScenarioRead
from app.services.scoring import weighted_score

router = APIRouter(prefix="/scenarios", tags=["scenarios"])


@router.get("/", response_model=list[ScenarioRead])
def list_scenarios(territory_id: int | None = None, db: Session = Depends(get_db)):
    from app.models.indicator import Scenario
    q = db.query(Scenario)
    if territory_id:
        q = q.filter(Scenario.territory_id == territory_id)
    return q.order_by(Scenario.id.desc()).all()


@router.post("/", response_model=ScenarioRead)
def create_scenario(payload: ScenarioCreate, db: Session = Depends(get_db)):
    from app.models.indicator import Scenario
    score = weighted_score(payload.parameters)
    sc = Scenario(territory_id=payload.territory_id, name=payload.name,
                  description=payload.description, parameters=payload.parameters,
                  results=score.breakdown, performance=score.global_score)
    db.add(sc); db.commit(); db.refresh(sc)
    return sc


@router.get("/{scenario_id}", response_model=ScenarioRead)
def get_scenario(scenario_id: int, db: Session = Depends(get_db)):
    from app.models.indicator import Scenario
    sc = db.get(Scenario, scenario_id)
    if not sc:
        raise HTTPException(404, "Scénario introuvable")
    return sc


@router.delete("/{scenario_id}")
def delete_scenario(scenario_id: int, db: Session = Depends(get_db)):
    from app.models.indicator import Scenario
    sc = db.get(Scenario, scenario_id)
    if not sc:
        raise HTTPException(404, "Scénario introuvable")
    db.delete(sc); db.commit()
    return {"deleted": scenario_id}

File: frontend/src/api/scenarios.ts
import { apiClient } from "./client";

export interface Scenario {
  id: number;
  territory_id: number;
  name: string;
  description?: string;
  parameters?: Record<string, number>;
  results?: Record<string, number>;
  performance?: number;
}

export interface ScenarioCreate {
  territory_id: number;
  name: string;
  description?: string;
  parameters: Record<string, number>;
}

export const fetchScenarios = async (territoryId?: number): Promise<Scenario[]> => {
  const { data } = await apiClient.get<Scenario[]>("/scenarios/", {
    params: territoryId ? { territory_id: territoryId } : {},
  });
  return data;
};

export const createScenario = async (payload: ScenarioCreate): Promise<Scenario> => {
  const { data } = await apiClient.post<Scenario>("/scenarios/", payload);
  return data;
};

export const deleteScenario = async (id: number): Promise<void> => {
  await apiClient.delete(`/scenarios/${id}`);
};
